### PR TITLE
Correct `exact_rows_before_limit` in some scenarios

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The parameter `exact_rows_before_limit` is used to make `rows_before_limit_at_least` is designed to accurately reflect the number of rows returned before the limit is reached. This pull request addresses issues encountered when the query involves distributed processing across multiple shards or sorting operations. Prior to this update, these scenarios were not functioning as intended.

(The above content is generated from GPT-3.5)

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Processors/IProcessor.h [hunk 1]
```diff
@@ -116,6 +117,7 @@ class IProcessor
 protected:
     InputPorts inputs;
     OutputPorts outputs;
+    RowsBeforeLimitCounterPtr rows_before_limit_at_least;
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/Processors/IProcessor.h:120]: Ok. Will add `setRowsBeforeLimitCounter` as virtual function only.

**Comment by KochetovNicolai** [comment]
Looks good generally

File: src/QueryPipeline/QueryPipeline.cpp [hunk 2]
```diff
@@ -129,84 +130,89 @@ static void checkCompleted(Processors & processors)
...
+            if (auto * sorting = typeid_cast<PartialSortingTransform *>(processor))
+            {
+                processors.emplace_back(processor);
+                continue;
+            }
+            else if (auto * having = typeid_cast<TotalsHavingTransform *>(processor))
+            {
+                /// Don't go to children if current transform changes the number of rows.
+                if (having->hasFilter())
+                    continue;
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/QueryPipeline/QueryPipeline.cpp]: Yes. I've also updated the comment. It should be much clearer.

File: tests/queries/0_stateless/01913_exact_rows_before_limit_full.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
+-- Tags: no-parallel, no-random-merge-tree-settings
+
+drop table if exists test;
+
+create table test (i int) engine MergeTree order by tuple();
+
+insert into test select arrayJoin(range(10000));
+
+set exact_rows_before_limit = 1, output_format_write_statistics = 0;
```
[the remark this answers is not in the document]
Comment (amosbird) [reply tests/queries/0_stateless/01913_exact_rows_before_limit_full.sql:9]: Done.

**Pull request merged**